spring-cloud-stream-binders/spring-cloud-stream-binder-gemfire/src/main/java/org/springframework/cloud/stream/binder/gemfire/GemfireMessageChannelBinder.java

```diff
@@ -0,0 +1,300 @@
+		logger.debug("bindReplier");
+	}
+
+
+	/**
+	 * {@link com.gemstone.gemfire.cache.CacheListener} implementation that
+	 * {@link Object#notifyAll() notifies} itself when a new entry is added
+	 * to the region it is registered for.
+	 */
+	private static class MessageListener extends CacheListenerAdapter<MessageKey, Message<?>> {
+
+		public MessageListener() {
+		}
+
+		@Override
+		public synchronized void afterCreate(EntryEvent<MessageKey, Message<?>> event) {
+			this.notifyAll();
+		}
+	}
+
+
+	/**
+	 * Reads {@link Message} objects from a {@link Region} and
+	 * publishes them to a {@link MessageChannel}.
+	 */
+	private static class QueueReader implements Runnable {
+
+		private final Region<MessageKey, Message<?>> messageRegion;
+
+		private final MessageChannel messageChannel;
+
+		private final MessageListener messageListener;
+
+		private volatile boolean running = true;
+
+		public QueueReader(Region<MessageKey, Message<?>> messageRegion,
+				MessageChannel messageChannel, MessageListener messageListener) {
+			this.messageRegion = messageRegion;
+			this.messageChannel = messageChannel;
+			this.messageListener = messageListener;
+		}
+
+		public boolean isRunning() {
+			return this.running;
+		}
+
+		public void shutdown() {
+			this.running = false;
+		}
+
+		@Override
+		public void run() {
+			// only messages that are present in this JVM will be processed
+			Region<MessageKey, Message<?>> localMessageRegion =
+					PartitionRegionHelper.getLocalData(this.messageRegion);
+
+			while (isRunning()) {
+				try {
+					List<MessageKey> keys = new ArrayList<>(localMessageRegion.keySet());
+					logger.debug("Fetched {} messages", keys.size());
+
+					if (!keys.isEmpty()) {
+						Collections.sort(keys);
+						List<MessageKey> errorKeys = null;
+						Map<MessageKey, Message<?>> messages = localMessageRegion.getAll(keys);
+						for (MessageKey key : keys) {
+							Message<?> message = messages.get(key);
+							logger.debug("QueueReader({})", message);
+							try {
+								this.messageChannel.send(message);
+							}
+							catch (Exception e) {
+								logger.warn("Exception processing message", e);
+								if (errorKeys == null) {
+									errorKeys = new ArrayList<>();
+								}
+								errorKeys.add(key);
+							}
+						}
+
+						// remove messages that were processed without error
+						// todo: consider adding un-processed messages to a "dead letter" region
+						if (errorKeys != null) {
+							keys.removeAll(errorKeys);
+						}
+						localMessageRegion.removeAll(keys);
+					}
+
+					synchronized (this.messageListener) {
+						this.messageListener.wait(500);
+					}
+				}
+				catch (InterruptedException e) {
+					logger.warn("Thread interrupted", e);
+					Thread.currentThread().interrupt();
+					shutdown();
+				}
+			}
+		}
+	}
+
+
+	/**
+	 * {@link MessageHandler} implementation that publishes messages
+	 * to a {@link Region}.
+	 */
+	private class SendingHandler implements MessageHandler {
```

Comment:
Please don't. I missed that reference, I'm absolutely fine with non-static inner class, if needed